<!DOCTYPE html PUBLIC "-//W3C//DTD XHTML 1.0 Transitional//EN" "https://www.w3.org/TR/xhtml1/DTD/xhtml1-transitional.dtd">
<html xmlns="http://www.w3.org/1999/xhtml" lang="en-US">
<head>
<meta http-equiv="Content-Type" content="text/xhtml;charset=UTF-8"/>
<meta http-equiv="X-UA-Compatible" content="IE=11"/>
<meta name="generator" content="Doxygen 1.12.0"/>
<meta name="viewport" content="width=device-width, initial-scale=1"/>
<title>scran_aggregate: Aggregate expression values across cells</title>
<link href="tabs.css" rel="stylesheet" type="text/css"/>
<script type="text/javascript" src="jquery.js"></script>
<script type="text/javascript" src="dynsections.js"></script>
<script type="text/javascript" src="clipboard.js"></script>
<link href="navtree.css" rel="stylesheet" type="text/css"/>
<script type="text/javascript" src="resize.js"></script>
<script type="text/javascript" src="cookie.js"></script>
<link href="search/search.css" rel="stylesheet" type="text/css"/>
<script type="text/javascript" src="search/searchdata.js"></script>
<script type="text/javascript" src="search/search.js"></script>
<script type="text/x-mathjax-config">
MathJax.Hub.Config({
  extensions: ["tex2jax.js"],
  jax: ["input/TeX","output/HTML-CSS"],
});
</script>
<script type="text/javascript" async="async" src="https://cdn.jsdelivr.net/npm/mathjax@2/MathJax.js"></script>
<link href="doxygen.css" rel="stylesheet" type="text/css" />
<link href="doxygen-awesome.css" rel="stylesheet" type="text/css"/>
</head>
<body>
<div id="top"><!-- do not remove this div, it is closed by doxygen! -->
<div id="titlearea">
<table cellspacing="0" cellpadding="0">
 <tbody>
 <tr id="projectrow">
  <td id="projectalign">
   <div id="projectname">scran_aggregate
   </div>
   <div id="projectbrief">Aggregate expression values across cells</div>
  </td>
 </tr>
 </tbody>
</table>
</div>
<!-- end header part -->
<!-- Generated by Doxygen 1.12.0 -->
<script type="text/javascript">
/* @license magnet:?xt=urn:btih:d3d9a9a6595521f9666a5e94cc830dab83b65699&amp;dn=expat.txt MIT */
var searchBox = new SearchBox("searchBox", "search/",'.html');
/* @license-end */
</script>
<script type="text/javascript">
/* @license magnet:?xt=urn:btih:d3d9a9a6595521f9666a5e94cc830dab83b65699&amp;dn=expat.txt MIT */
$(function() { codefold.init(0); });
/* @license-end */
</script>
<script type="text/javascript" src="menudata.js"></script>
<script type="text/javascript" src="menu.js"></script>
<script type="text/javascript">
/* @license magnet:?xt=urn:btih:d3d9a9a6595521f9666a5e94cc830dab83b65699&amp;dn=expat.txt MIT */
$(function() {
  initMenu('',true,false,'search.php','Search',false);
  $(function() { init_search(); });
});
/* @license-end */
</script>
<div id="main-nav"></div>
</div><!-- top -->
<div id="doc-content">
<script type="text/javascript">
/* @license magnet:?xt=urn:btih:d3d9a9a6595521f9666a5e94cc830dab83b65699&amp;dn=expat.txt MIT */
$(function(){ initResizable(false); });
/* @license-end */
</script>
<!-- window showing the filter options -->
<div id="MSearchSelectWindow"
     onmouseover="return searchBox.OnSearchSelectShow()"
     onmouseout="return searchBox.OnSearchSelectHide()"
     onkeydown="return searchBox.OnSearchSelectKey(event)">
</div>

<!-- iframe showing the search results (closed by default) -->
<div id="MSearchResultsWindow">
<div id="MSearchResults">
<div class="SRPage">
<div id="SRIndex">
<div id="SRResults"></div>
<div class="SRStatus" id="Loading">Loading...</div>
<div class="SRStatus" id="Searching">Searching...</div>
<div class="SRStatus" id="NoMatches">No Matches</div>
</div>
</div>
</div>
</div>

<div><div class="header">
  <div class="headertitle"><div class="title">Aggregate expression values across cells </div></div>
</div><!--header-->
<div class="contents">
<div class="textblock"><p><a class="anchor" id="md__2github_2workspace_2README"></a></p>
<p><img src="https://github.com/libscran/scran_aggregate/actions/workflows/run-tests.yaml/badge.svg" alt="Unit tests" style="pointer-events: none;" class="inline"/> <img src="https://github.com/libscran/scran_aggregate/actions/workflows/doxygenate.yaml/badge.svg" alt="Documentation" style="pointer-events: none;" class="inline"/> <a href="https://codecov.io/gh/libscran/scran_aggregate"><img src="https://codecov.io/gh/libscran/scran_aggregate/graph/badge.svg?token=JWV0I4WJX2" alt="Codecov" style="pointer-events: none;" class="inline"/></a></p>
<h1>Overview</h1>
<p>This repository contains functions to aggregate statistics for groups of cells or sets of genes from a gene-by-cell matrix of expression values. It was primarily developed for computing pseudo-bulk expression profiles for clusters of cells, which can then be used for differential expression analysis with packages like <b>edgeR</b>. The code itself was originally derived from the <a href="https://bioconductor.org/packages/scran"><b>scran</b> R package</a>, factored out into a separate C++ library for easier re-use.</p>
<h1>Quick start</h1>
<p>Given a <a href="https://github.com/tatami-inc/tatami"><code>tatami::Matrix</code></a> and an array of group assignments, the <code>aggregate_across_cells()</code> function will compute the aggregate statistics across all genes for each group.</p>
<div class="fragment"><div class="line"><span class="preprocessor">#include &quot;<a class="code" href="scran__aggregate_8hpp.html">scran_aggregate/scran_aggregate.hpp</a>&quot;</span></div>
<div class="line"> </div>
<div class="line"><span class="keyword">const</span> <a class="code hl_classRef" href="https://tatami-inc.github.io/tatami/classtatami_1_1Matrix.html">tatami::Matrix&lt;double, int&gt;</a>&amp; mat = some_data_source();</div>
<div class="line"> </div>
<div class="line"><span class="comment">// Array of groupings should contain integer assignments to groups 0, 1, 2, etc.</span></div>
<div class="line">std::vector&lt;int&gt; groupings = some_groupings();</div>
<div class="line"> </div>
<div class="line"><a class="code hl_struct" href="structscran__aggregate_1_1AggregateAcrossCellsOptions.html">scran_aggregate::AggregateAcrossCellsOptions</a> opt;</div>
<div class="line"><span class="keyword">auto</span> res = <a class="code hl_function" href="namespacescran__aggregate.html#a455854154026f84c3d8c52a216472c2b">scran_aggregate::aggregate_across_cells</a>(mat, groupings.data(), opt);</div>
<div class="line"> </div>
<div class="line">res.sums; <span class="comment">// vector of vectors of per-group sums across genes.</span></div>
<div class="line">res.sums[0]; <span class="comment">// vector of sums for the first group across genes.</span></div>
<div class="line">res.detected; <span class="comment">// vector of vectors of the number of detected cells per gene.</span></div>
<div class="ttc" id="aclasstatami_1_1Matrix_html"><div class="ttname"><a href="https://tatami-inc.github.io/tatami/classtatami_1_1Matrix.html">tatami::Matrix</a></div></div>
<div class="ttc" id="anamespacescran__aggregate_html_a455854154026f84c3d8c52a216472c2b"><div class="ttname"><a href="namespacescran__aggregate.html#a455854154026f84c3d8c52a216472c2b">scran_aggregate::aggregate_across_cells</a></div><div class="ttdeci">void aggregate_across_cells(const tatami::Matrix&lt; Data_, Index_ &gt; &amp;input, const Group_ *const group, const AggregateAcrossCellsBuffers&lt; Sum_, Detected_ &gt; &amp;buffers, const AggregateAcrossCellsOptions &amp;options)</div><div class="ttdef"><b>Definition</b> aggregate_across_cells.hpp:273</div></div>
<div class="ttc" id="ascran__aggregate_8hpp_html"><div class="ttname"><a href="scran__aggregate_8hpp.html">scran_aggregate.hpp</a></div><div class="ttdoc">Aggregate single-cell expression values.</div></div>
<div class="ttc" id="astructscran__aggregate_1_1AggregateAcrossCellsOptions_html"><div class="ttname"><a href="structscran__aggregate_1_1AggregateAcrossCellsOptions.html">scran_aggregate::AggregateAcrossCellsOptions</a></div><div class="ttdoc">Options for aggregate_across_cells().</div><div class="ttdef"><b>Definition</b> aggregate_across_cells.hpp:25</div></div>
</div><!-- fragment --><p>We can also use the <code>aggregate_across_genes()</code> function to sum expression values across gene sets, e.g., to compute the activity of a gene signature. This can be done with any number of gene sets, possibly with a different weight for each gene in each set.</p>
<div class="fragment"><div class="line">std::vector&lt;std::tuple&lt;size_t, const int*, const double*&gt; &gt; gene_sets;</div>
<div class="line"> </div>
<div class="line">std::vector&lt;int&gt; set1 { 0, 5, 10, 20 };</div>
<div class="line">gene_sets.emplace_back(set1.size(), set1.data(), <span class="keyword">static_cast&lt;</span><span class="keywordtype">double</span>*<span class="keyword">&gt;</span>(NULL)); <span class="comment">// no weight</span></div>
<div class="line"> </div>
<div class="line">std::vector&lt;int&gt; set2 { 0, 2, 4, 6, 8, 10 };</div>
<div class="line">std::vector&lt;double&gt; weight2 { 0.1, 0.3, 0.3, 0.2, 0.1, 0.05 };</div>
<div class="line">gene_sets.emplace_back(set2.size(), set2.data(), weight2.data()); <span class="comment">// weighted</span></div>
<div class="line"> </div>
<div class="line"><a class="code hl_struct" href="structscran__aggregate_1_1AggregateAcrossGenesOptions.html">scran_aggregate::AggregateAcrossGenesOptions</a> g_opt;</div>
<div class="line"><span class="keyword">auto</span> g_res = <a class="code hl_function" href="namespacescran__aggregate.html#abf6732ea8012e17cc65039228663eae0">scran_aggregate::aggregate_across_genes</a>(</div>
<div class="line">    mat,</div>
<div class="line">    gene_sets,</div>
<div class="line">    g_opt</div>
<div class="line">);</div>
<div class="line">g_res.sum[0]; <span class="comment">// vector of sums for set 1 in each cell.</span></div>
<div class="ttc" id="anamespacescran__aggregate_html_abf6732ea8012e17cc65039228663eae0"><div class="ttname"><a href="namespacescran__aggregate.html#abf6732ea8012e17cc65039228663eae0">scran_aggregate::aggregate_across_genes</a></div><div class="ttdeci">void aggregate_across_genes(const tatami::Matrix&lt; Data_, Index_ &gt; &amp;input, const std::vector&lt; std::tuple&lt; std::size_t, const Gene_ *, const Weight_ * &gt; &gt; &amp;gene_sets, const AggregateAcrossGenesBuffers&lt; Sum_ &gt; &amp;buffers, const AggregateAcrossGenesOptions &amp;options)</div><div class="ttdef"><b>Definition</b> aggregate_across_genes.hpp:274</div></div>
<div class="ttc" id="astructscran__aggregate_1_1AggregateAcrossGenesOptions_html"><div class="ttname"><a href="structscran__aggregate_1_1AggregateAcrossGenesOptions.html">scran_aggregate::AggregateAcrossGenesOptions</a></div><div class="ttdoc">Options for aggregate_across_genes().</div><div class="ttdef"><b>Definition</b> aggregate_across_genes.hpp:26</div></div>
</div><!-- fragment --><p>Check out the <a href="https://libscran.github.io/scran_aggregate">reference documentation</a> for more details.</p>
<h1>Building projects</h1>
<h2>CMake with <code>FetchContent</code></h2>
<p>If you're using CMake, you just need to add something like this to your <code>CMakeLists.txt</code>:</p>
<div class="fragment"><div class="line">include(FetchContent)</div>
<div class="line"> </div>
<div class="line">FetchContent_Declare(</div>
<div class="line">  scran_aggregate</div>
<div class="line">  GIT_REPOSITORY https://github.com/libscran/scran_aggregate</div>
<div class="line">  GIT_TAG master # or any version of interest</div>
<div class="line">)</div>
<div class="line"> </div>
<div class="line">FetchContent_MakeAvailable(scran_aggregate)</div>
</div><!-- fragment --><p>Then you can link to <b><a class="el" href="namespacescran__aggregate.html" title="Aggregate single-cell expression values.">scran_aggregate</a></b> to make the headers available during compilation:</p>
<div class="fragment"><div class="line"># For executables:</div>
<div class="line">target_link_libraries(myexe libscran::scran_aggregate)</div>
<div class="line"> </div>
<div class="line"># For libaries</div>
<div class="line">target_link_libraries(mylib INTERFACE libscran::scran_aggregate)</div>
</div><!-- fragment --><h2>CMake with <code>find_package()</code></h2>
<div class="fragment"><div class="line">find_package(libscran_scran_aggregate CONFIG REQUIRED)</div>
<div class="line">target_link_libraries(mylib INTERFACE libscran::scran_aggregate)</div>
</div><!-- fragment --><p>To install the library, use:</p>
<div class="fragment"><div class="line">mkdir build &amp;&amp; cd build</div>
<div class="line">cmake .. -DSCRAN_AGGREGATE_TESTS=OFF</div>
<div class="line">cmake --build . --target install</div>
</div><!-- fragment --><p>By default, this will use <code>FetchContent</code> to fetch all external dependencies. If you want to install them manually, use <code>-DSCRAN_AGGREGATE_FETCH_EXTERN=OFF</code>. See the tags in <a href="extern/CMakeLists.txt"><code>extern/CMakeLists.txt</code></a> to find compatible versions of each dependency.</p>
<h2>Manual</h2>
<p>If you're not using CMake, the simple approach is to just copy the files in <code>include/</code> - either directly or with Git submodules - and include their path during compilation with, e.g., GCC's <code>-I</code>. This also requires the external dependencies listed in <a href="extern/CMakeLists.txt"><code>extern/CMakeLists.txt</code></a>. </p>
</div></div><!-- PageDoc -->
<a href="doxygen_crawl.html"></a>
</div><!-- contents -->
<!-- start footer part -->
<hr class="footer"/><address class="footer"><small>
Generated by&#160;<a href="https://www.doxygen.org/index.html"><img class="footer" src="doxygen.svg" width="104" height="31" alt="doxygen"/></a> 1.12.0
</small></address>
</div><!-- doc-content -->
</body>
</html>
